File: 206_reverse_list.cpp
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        // iterative
        if (head == nullptr) return head;
        ListNode* next;
        ListNode* prev = nullptr;
        while(head) {
            next = head->next;
            head->next = prev;
            prev = head;
            head = next;
        }
        return prev;

        // Recursion (ListNode* prev = nullptr parameter)
        // if (!head) return prev;
        // ListNode* next = head->next;
        // head->next = prev;
        // return reverseList(next, head);

        // Recursion one-liner (two parameters)
        // return head ? reverseList(head->next, (head->next = prev, head)) : prev; 
    }
};

File: 206_reverse_list.java
class Solution {
    public ListNode reverseList(ListNode head) {
        // // rewrite of C++
        // if (head == null) { return head; }
        // ListNode next;
        // ListNode prev = null; 
        // while (head != null) {
        //     next = head.next;
        //     head.next = prev;
        //     prev = head;
        //     head = next;
        // }
        // return prev;

        // recursive 
        if (head == null || head.next == null) { return head; }
        ListNode tmp = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return tmp;
    }
}
